webrtc-c/canary/src/Main.cpp

```diff
@@ -86,11 +100,20 @@ STATUS run(Canary::PConfig pConfig)
         CHK_STATUS(peer.init());
         CHK_STATUS(peer.connect());
 
-        std::thread videoThread(sendLocalFrames, &peer, MEDIA_STREAM_TRACK_KIND_VIDEO, "./assets/h264SampleFrames/frame-%04d.h264",
-                                NUMBER_OF_H264_FRAME_FILES, SAMPLE_VIDEO_FRAME_DURATION);
-        std::thread audioThread(sendLocalFrames, &peer, MEDIA_STREAM_TRACK_KIND_AUDIO, "./assets/opusSampleFrames/sample-%03d.opus",
-                                NUMBER_OF_OPUS_FRAME_FILES, SAMPLE_AUDIO_FRAME_DURATION);
-
+        std::thread videoThread(sendLocalFrames, &peer, MEDIA_STREAM_TRACK_KIND_VIDEO, CANARY_VIDEO_FRAMES_PATH, NUMBER_OF_H264_FRAME_FILES,
+                                SAMPLE_VIDEO_FRAME_DURATION);
+        std::thread audioThread(sendLocalFrames, &peer, MEDIA_STREAM_TRACK_KIND_AUDIO, CANARY_AUDIO_FRAMES_PATH, NUMBER_OF_OPUS_FRAME_FILES,
+                                SAMPLE_AUDIO_FRAME_DURATION);
+
+        // All metrics tracking will happen on a time queue to simplify handling periodicity
+        CHK_STATUS(timerQueueAddTimer(timerQueueHandle, METRICS_INVOCATION_PERIOD, METRICS_INVOCATION_PERIOD, canaryRtpOutboundStats, (UINT64) &peer,
+                                      &timeoutTimerId));
+        CHK_STATUS(timerQueueAddTimer(timerQueueHandle, METRICS_INVOCATION_PERIOD, METRICS_INVOCATION_PERIOD, canaryRtpInboundStats, (UINT64) &peer,
+                                      &timeoutTimerId));
+        if (!pConfig->isMaster) {
+            CHK_STATUS(timerQueueAddTimer(timerQueueHandle, END_TO_END_METRICS_INVOCATION_PERIOD, END_TO_END_METRICS_INVOCATION_PERIOD,
+                                          canaryEndToEndStats, (UINT64) &peer, &timeoutTimerId));
+        }
         videoThread.join();
         audioThread.join();
         CHK_STATUS(peer.shutdown());
```

Comment:
Important: the metric timers need to be stopped before shutting down peer. Otherwise, it'll try to use peer that has been shutdown and might cause a crash.

Probably guard each callback by `terminated.load()`. If it's terminated, you can return the timer cancellation status.

Reply:
Good point. Done.